src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsAllShardsMetricsCollector.java

```diff
@@ -100,25 +109,28 @@ public void collectMetrics(long startTime) {
 
         try {
             populateCurrentShards();
-            // Metrics populated for all shards in every collection.
-            for (HashMap.Entry currentShard : currentShards.entrySet() ){
-                IndexShard currentIndexShard = (IndexShard)currentShard.getValue();
-                IndexShardStats currentIndexShardStats = Utils.indexShardStats(indicesService,
-                        currentIndexShard, new CommonStatsFlags(CommonStatsFlags.Flag.QueryCache,
-                                CommonStatsFlags.Flag.FieldData,
-                                CommonStatsFlags.Flag.RequestCache));
-                for (ShardStats shardStats : currentIndexShardStats.getShards()) {
-                    StringBuilder value = new StringBuilder();
-
-                    value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds());
-                    // Populate the result with cache specific metrics only.
-                    value.append(PerformanceAnalyzerMetrics.sMetricNewLineDelimitor)
-                            .append(new NodeStatsMetricsAllShardsPerCollectionStatus(shardStats).serialize());
-                    saveMetricValues(value.toString(), startTime, currentIndexShardStats.getShardId().getIndexName(),
-                            String.valueOf(currentIndexShardStats.getShardId().id()));
+            populatePerShardStats(indicesService);
+
+            for (HashMap.Entry currentShard : currentPerShardStats.entrySet()) {
+                ShardId shardId = (ShardId) currentShard.getKey();
+                ShardStats currentShardStats = (ShardStats) currentShard.getValue();
+                if (prevPerShardStats.size() == 0) {
+                    // Populating value for the first run.
+                    populateMetricValue(currentShardStats, startTime, shardId.getIndexName(), shardId.id());
+                    continue;
+                }
+                ShardStats prevShardStats = prevPerShardStats.get(shardId);
+                if (prevShardStats != null) {
+                    // Populate value for shards which are new and were not present in the previous run.
+                    populateMetricValue(currentShardStats, startTime, shardId.getIndexName(), shardId.id());
+                    continue;
                 }
+                NodeStatsMetricsAllShardsPerCollectionStatus prevValue = new
+                        NodeStatsMetricsAllShardsPerCollectionStatus(prevShardStats);
+                NodeStatsMetricsAllShardsPerCollectionStatus currValue = new
+                        NodeStatsMetricsAllShardsPerCollectionStatus(currentShardStats);
+                populateDiffMetricValue(prevValue, currValue, startTime, shardId.getIndexName(), shardId.id());
```

Comment:
Instead of creating the `NodeStatsMetricsAllShardsPerCollectionStatus` object, can we instead just use the `prevShardStats` and `currentShardStats` in `populateDiffMetricValue`

Reply:
We cant do that directly from the ShardStats as the values for different metrics are not stored as variables of the ShardStats class but those are fetched via different getters individually.